internlm/monitor/diagnosis.py

```diff
@@ -0,0 +1,272 @@
+import os
+import torch
+import socket
+import functools
+import contextlib
+from internlm.utils.registry import MODEL_INITIALIZER
+from internlm.utils.writer import Writer
+from internlm.utils.megatron_timers import megatron_timer
+from internlm.utils.megatron_timers import megatron_timer as timer
+from typing import Callable, Iterable, Union
+from internlm.utils.logger import get_logger, initialize_uniscale_logger
+from internlm.core.context import ParallelMode
+from internlm.core.context import global_context as gpc
+from internlm.utils.parallel import get_parallel_log_file_name
+from internlm.utils.common import DummyProfile
+from internlm.utils.gputest import bench_gpu, bench_net
+from beautifultable import BeautifulTable
+from collections import defaultdict
+import torch.distributed as dist
+
+
+logger = get_logger(__file__)
+
+# def initialize_llm_logger(start_time: str, do_diagnosis=False):
+#     """
+#     Initialize customed uniscale logger.
+
+#     Args:
+#         start_time (str): The launch time of current training job.
+
+#     Returns: The instance of uniscale logger.
+#     """
+#     print('do_diagnosis', do_diagnosis, flush=True)
+#     uniscale_logger = initialize_uniscale_logger(
+#         job_name=gpc.config.JOB_NAME, launch_time=start_time, file_name=get_parallel_log_file_name(), do_diagnosis=do_diagnosis
+#     )
+#     # if uniscale_logger is not None:
+#     #     global logger
+#     #     logger = uniscale_logger
+
+#     return uniscale_logger
+
+class LLMProfiler:
+    def __init__(
+        self,
+        timer: megatron_timer,
+        train_state,
+        log_time_step: int = 1,
+        launch_time: str = None,
+        active_count: int = 1,
+        do_trace_profiling: bool = False,
+        trace_profiling_range: Callable = None,
+        do_diagnosis: bool = False,
+        writer: Writer = None,
+        current_time = None,
+    ) -> None:
+        """
+        TODO: support PP diagnosis mode.
+
+        Args:
+            timer (megatron_timer): 
+            train_state (_type_): 
+            log_time_step (int, optional): . Defaults to 1.
+            trace_profiling_range (Callable, optional): A user-defined function with a return value of true or false, 
+                used to control whether trace profiling is enabled. Defaults to None.
+            launch_time (str, optional): . Defaults to None.
+            active_count (int, optional): trace profiling interval. Defaults to 1.
+            do_trace_profiling (bool): Whether to enable trace profiling. Defaults to False.
+            do_diagnosis (bool, optional): Whether to enable runtime diagnostics. Defaults to False.
+            writer (Writer, optional): . Defaults to None.
+        """
+        from datetime import datetime
+
+        self.trace_profiling = do_trace_profiling
+        self.train_state = train_state
+        self.active_count = active_count
```

Comment:
we have three profiler(torch.profiler, simple_memory_profiler, diagnosis profiler) they all need to increments step counter and are dependent on different `active_count`. Maybe add a new config dict in ./configs, it would be clearer.